lib/core/utils/preload-cssom.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * Returns a then(able) queue of CSSStyleSheet(s)
+ * @param {Object} ownerDocument document object to be inspected for stylesheets
+ * @param {number} timeout on network request for stylesheet that need to be externally fetched
+ * @param {Function} getSheetFromTextFn a utility function to generate a style sheet from text
+ * @return {Object} queue
+ * @private
+ */
+function loadCssom(ownerDocument, timeout, getSheetFromTextFn) {
+	const q = axe.utils.queue();
+
+	Array.from(ownerDocument.styleSheets).forEach(sheet => {
+		if (sheet.disabled) {
+			return;
+		}
+
+		try {
+			if (sheet.cssRules) {
+				// accessing .cssRules throws for external (cross-domain) sheets, which is handled in the catch
+				q.defer(resolve => resolve(sheet));
+			}
+		} catch (e) {
+			const deferredSheet = (resolve, reject) => {
+				axe.imports
+					.axios({
+						method: 'get',
+						url: sheet.href,
+						timeout
+					})
+					.then(({ data }) => {
+						const sheet = getSheetFromTextFn(data);
+						resolve(sheet);
+					})
+					.catch(reject);
+			};
+			// external sheet -> make an xhr and q the response
+			q.defer(deferredSheet);
+		}
+	});
+
+	return q;
+}
+
+/**
+ * Returns an array of documents with a given root node/ tree
+ * @param {Object} treeRoot - the DOM tree to be inspected
+ * @return {Array} documents
+ * @private
+ */
+function getDocumentsFromTreeRoot(treeRoot) {
+	let ids = [];
+	const documents = axe.utils
+		.querySelectorAllFilter(treeRoot, '*', node => {
+			if (ids.includes(node.shadowId)) {
```

Comment:
since shadowId can be undefined, might be safer if we test that explicitly, so `node.shadowId && ids.includes(...`. We've been burned by bad polyfills before.

Reply:
I understand the thinking here, but node.shadowId seems to be undefined 13 times in my example from the qsa function. A bit confusing, can we chat on this?

Reply:
That is cool by me 🍺